### PR TITLE
Implemented things required for octo compatibility

### PR DESCRIPTION
**Pull request by susch19**
* new deserialize method with out parameters of every serialized property
* new deserialize method with instance as an input
* reworked system to allow generation of additional code / methods inside of serializers

### CODE REVIEW

File: DEMO/SUTMessage.cs [hunk 1]
```diff
@@ -199,7 +201,7 @@ public override int GetHashCode()
             return hash.ToHashCode();
         }
 
-        [NoosonCustom(SerializeMethodName = "SerializeMe", DeserializeMethodName = "DeserializeMe")]
+        [NoosonCustom(SerializeMethodName = "SerializeMe")]
```
Comment (jvbsl) [inline DEMO/SUTMessage.cs:204]: Why was that removed?

File: DEMO/SerializeCallTest.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+{
+    public int MyProperty { get; set; }
+    public static OverrideSerializeCalLTest Deserialize(BinaryReader br)
+    {
+        throw new NotImplementedException();
+    }
+    //public override void Deserialize(BinaryReader br)
+    //{
+    //    throw new NotImplementedException();
+    //}
```
Comment (jvbsl) [inline DEMO/SerializeCallTest.cs:28]:
```suggestion
```

File: DEMO/SerializeCallTest.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+public abstract class BaseSerializeCallTest
+{
+    public int MyBaseProp { get; set; }
+
+    public virtual void Serialize(BinaryWriter bw)
+    {
+    }
+    //public abstract void Deserialize(BinaryReader br);
```
Comment (jvbsl) [inline DEMO/SerializeCallTest.cs:43]:
```suggestion
```

File: DEMO/StructTestFile.cs [hunk 4]
```diff
@@ -1,9 +1,11 @@
 using NonSucking.Framework.Serialization;
 
+using System.Numerics;
+
 namespace DEMO;
 
 [Nooson]
-public partial struct TestStruct
+public partial struct TestStruct
```
Comment (jvbsl) [inline DEMO/StructTestFile.cs]:
```suggestion
```

File: DEMO/StructTestFile.cs [hunk 5]
```diff
@@ -1,9 +1,11 @@
 using NonSucking.Framework.Serialization;
 
+using System.Numerics;
+
```
Comment (jvbsl) [inline DEMO/StructTestFile.cs]:
```suggestion
```

File: DEMO/Versioning.cs [hunk 6]
```diff
@@ -22,9 +23,73 @@ public partial class Versioning
...
+    //        long NewProp3__ret__uG = default(long)!;
+    //        if (Checker(Version__ret__uC, NewProp__ret__uE))
+    //        {
+    //            NewProp3__ret__uG = reader.ReadInt64();
+    //        }
+    //        else
+    //            NewProp3__ret__uG = 123;
+    //    }
+    //    ... = variable;
+    //}
```
Comment (jvbsl) [inline DEMO/Versioning.cs]:
```suggestion
```

File: DEMO/Versioning.cs [hunk 7]
```diff
@@ -1,5 +1,6 @@
-﻿using Microsoft.CodeAnalysis.CSharp.Syntax;
-using NonSucking.Framework.Serialization;
+﻿using NonSucking.Framework.Serialization;
+
+using System.Text.Json.Serialization;
```
Comment (jvbsl) [inline DEMO/Versioning.cs]:
```suggestion
```

File: DEMO/Versioning.cs [hunk 8]
```diff
@@ -22,9 +23,73 @@ public partial class Versioning
     [NoosonVersioning(nameof(Checker), "123", nameof(Version), nameof(NewProp))]
     public long NewProp3 { get; set; }
 
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="version">The abc <see cref="Version"/></param>
+    /// <param name="another"></param>
+    /// <returns></returns>
```
Comment (jvbsl) [inline DEMO/Versioning.cs]:
```suggestion
```

File: NonSucking.Framework.Serialization/Serializers/MethodCallSerializer.cs [hunk 9]
```diff
@@ -17,18 +17,18 @@ namespace NonSucking.Framework.Serialization
     [StaticSerializer(40)]
     internal static class MethodCallSerializer
     {
-        internal static bool TryDeserialize(MemberInfo property, NoosonGeneratorContext context, string readerName, GeneratedSerializerCode statements, SerializerMask includedSerializers)
+        internal static bool TryDeserialize( MemberInfo property, NoosonGeneratorContext context, string readerName, GeneratedSerializerCode statements, SerializerMask includedSerializers)
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/MethodCallSerializer.cs]:
```suggestion
        internal static bool TryDeserialize(MemberInfo property, NoosonGeneratorContext context, string readerName, GeneratedSerializerCode statements, SerializerMask includedSerializers)
```

File: NonSucking.Framework.Serialization/GeneratedMethod.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            if (other is null)
+                return false;
+            if (ReferenceEquals(this, other))
+                return true;
+            return Name == other.Name
+                && (Parameters.Equals(other.Parameters)
+                    || Parameters.SequenceEqual(other.Parameters))
+                && (Modifier.Equals(other.Modifier)
+                    || Modifier.SequenceEqual(other.Modifier));
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/GeneratedMethod.cs]:
```suggestion
                    || Modifier.SequenceEqual(other.Modifier));
                && (TypeParameters.Equals(other.TypeParameters)
                    || TypeParameters.SequenceEqual(other.TypeParameters));
```

File: NonSucking.Framework.Serialization/Helper.cs [hunk 11]
```diff
@@ -78,14 +104,15 @@ internal static string GetReadMethodCallFrom(SpecialType specialType)
                 _ => throw new NotSupportedException(),
             };
         }
-
+
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Helper.cs]:
```suggestion

```

File: NonSucking.Framework.Serialization/NoosonGenerator.cs [hunk 12]
```diff
@@ -267,327 +197,494 @@ private static string TypeNameToSummaryName(string typeName)
...
+            }
+
             var body
-                = CreateBlock(member, context, MethodType.Deserialize);
-        
-            var modifiers = new List<Modifiers> { Modifiers.Public };
-        
+                = CreateBlock(member, context, methodType);
+
+            var modifiers = new List<Modifiers> { Modifiers.Public, Modifiers.Static };
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/NoosonGenerator.cs:637]: Don't we want the context modifiers here instead of hardcoded public(as a copy)?

File: NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs [hunk 13]
```diff
@@ -13,9 +13,72 @@
...
+    [Flags]
+    internal enum Initializer
+    {
+        None = 0,
+        Ctor = 1 << 1,
+        /// <summary>
+        /// Will enable <see cref="Ctor"/> aswell
+        /// </summary>
+        InitializerList = (1 << 2) | Ctor,
+        Properties = 1 << 3,
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs]:
```suggestion
        Ctor = 1 << 0,
        /// <summary>
        /// Will enable <see cref="Ctor"/> aswell
        /// </summary>
        InitializerList = (1 << 1) | Ctor,
        Properties = 1 << 2,
```
No reason to start at 2?

File: NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs [hunk 14]
```diff
@@ -13,9 +13,72 @@
...
+    }
+
     internal static class CtorSerializer
     {
-        internal static GeneratedSerializerCode CallCtorAndSetProps(INamedTypeSymbol typeSymbol, ICollection<StatementSyntax> statements, MemberInfo instance, string instanceName)
+        internal static GeneratedSerializerCode CallCtorAndSetProps(INamedTypeSymbol typeSymbol, List<string> statements, MemberInfo instance, string instanceName, Initializer initializer)
+        {
+            List<string> localDeclarations = GetLocalDeclarations(statements, instanceName);
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs]:
```suggestion
        internal static GeneratedSerializerCode CallCtorAndSetProps(INamedTypeSymbol typeSymbol, List<string> localVariableNames, MemberInfo instance, string instanceName, Initializer initializer)
        {
            List<string> localDeclarations = GetLocalDeclarations(localVariableNames, instanceName);
```

File: NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs [hunk 15]
```diff
@@ -24,19 +87,18 @@ var constructors
...
-            var shouldContain = instanceName.Substring(0, instanceName.IndexOf(Helper.localVariableSuffix));
+        private static List<string> GetLocalDeclarations(List<string> statements, string instanceName)
+        {
+            var indexOf = instanceName.IndexOf(Helper.localVariableSuffix);
+            if (indexOf < 0)
+                return statements;
+            var shouldContain = instanceName.Substring(0, indexOf);
 
             var localDeclarations
                 = statements
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs]:
```suggestion
        private static List<string> GetLocalDeclarations(List<string> localVariableNames, string instanceName)
        {
            var indexOf = instanceName.IndexOf(Helper.localVariableSuffix);
            if (indexOf < 0)
                return localVariableNames;
            var shouldContain = instanceName.Substring(0, indexOf);

            var localDeclarations
                = localVariableNames
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 16]
```diff
@@ -17,132 +21,528 @@ namespace NonSucking.Framework.Serialization
...
+            (List<(string typeName, string parameterName)> Parameters, string typeName, IMethodSymbol? Symbol)? hasBaseDeserialize)
         {
-            var props
-               = Helper.GetMembersWithBase(property.TypeSymbol, baseTypesLevelProperties)
-               .Where(property =>
-                   property.Name != "this[]")
-               .Select(x => x with { Parent = property.Name });
+            if (property.Symbol.Name == "ListsContainer")
+                ;
+
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
```

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 17]
```diff
@@ -17,132 +21,528 @@ namespace NonSucking.Framework.Serialization
...
+            IEnumerable<GeneratedMethodParameter> leadingMethodParameters,
+            (List<(string typeName, string parameterName)> Parameters, string typeName, IMethodSymbol? Symbol)? hasBaseDeserialize,
+            Dictionary<string, string> scopeVariableNameMappings,
+            List<MemberInfo> filteredProps,
+            List<(bool assign, GeneratedSerializerCode propCode)> propCodes,
+            ITypeSymbol typeSymbol,
+            GeneratedType genType)
+        {
+            GeneratedMethod? genMethod =
+                genType.Methods.FirstOrDefault(x => x.Name == Consts.Deserialize /*&& x.Typestuff*/);
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
                genType.Methods.FirstOrDefault(x => x.Name == Consts.Deserialize /* TODO: && x.Typestuff*/);
```
so that we can find it again atleast

File: NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs [hunk 18]
```diff
@@ -17,132 +21,528 @@ namespace NonSucking.Framework.Serialization
...
+                    {
+                        genMethod.Body.Statements.Add(Statement.Declaration.Assign(outVarName, valueToAssign));
+
+                    }
+                    genMethod.Body.Statements.AddRange(propCode.Statements);
                 }
-                statements.Statements.AddRange(propCode.ToMergedBlock());
             }
+            else
+                ;
```
Comment (jvbsl) [inline NonSucking.Framework.Serialization/Serializers/PublicPropertySerializer.cs]:
```suggestion
```

**Review by jvbsl — CHANGES_REQUESTED** [review]

File: DEMO/SUTMessage.cs (hunk 1 above)
Comment (susch19) [reply DEMO/SUTMessage.cs:204]: Possibly for testing purposes

File: DEMO/SerializeCallTest.cs (hunk 2 above)
Comment (susch19) [reply DEMO/SerializeCallTest.cs:28]: This is most likely for manually adding and removing for testing the calls

File: DEMO/SerializeCallTest.cs (hunk 3 above)
Comment (susch19) [reply DEMO/SerializeCallTest.cs:43]: This is most likely for manually adding and removing for testing the calls

File: NonSucking.Framework.Serialization/GeneratedMethod.cs (hunk 10 above)
Comment (susch19) [reply NonSucking.Framework.Serialization/GeneratedMethod.cs]: That would be an invalid syntax, but i know what you mean

File: NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs (hunk 13 above)
Comment (susch19) [reply NonSucking.Framework.Serialization/Serializers/CtorSerializer.cs]: Accidentically done it

File: NonSucking.Framework.Serialization/NoosonGenerator.cs (hunk 12 above)
Comment (susch19) [reply NonSucking.Framework.Serialization/NoosonGenerator.cs:637]: No, because the context only has public as a modifier, but deserialize is (currently) also public static and override will never be used, so it can only be these 3 modifiers (maybe for now)

**Review by jvbsl — APPROVED** [review]
